lahja/trio/endpoint.py

```diff
@@ -0,0 +1,905 @@
+        :class:`~lahja.common.BroadcastConfig` to decide where the request
+        should be broadcasted to. By default, requests are broadcasted across
+        all connected endpoints with their consuming call sites.
+        """
+        request_id = next(self._get_request_id)
+
+        # Create an asynchronous generator that we use to pipe the result
+        send_channel, receive_channel = cast(
+            RequestResponseChannelPair, trio.open_memory_channel(0)
+        )
+
+        # place the send channel where the message processing loop can find it.
+        self._pending_requests[request_id] = send_channel
+
+        await self._broadcast(item, config, request_id)
+
+        # await for the result to be sent through the channel.
+        result = await receive_channel.receive()
+        await send_channel.aclose()
+        expected_response_type = item.expected_response_type()
+        if not isinstance(result, expected_response_type):
+            raise UnexpectedResponse(
+                f"The type of the response is {type(result)}, expected: "
+                f"{expected_response_type}"
+            )
+
+        return result
+
+    TSubscribeEvent = TypeVar("TSubscribeEvent", bound=BaseEvent)
+
+    def subscribe(
+        self,
+        event_type: Type[TSubscribeEvent],
+        handler: Callable[[TSubscribeEvent], Union[Any, Awaitable[Any]]],
+    ) -> Subscription:
+        """
+        Subscribe to receive updates for any event that matches the specified event type.
+        A handler is passed as a second argument an :class:`~lahja.common.Subscription` is returned
+        to unsubscribe from the event if needed.
+        """
+        if asyncio.iscoroutinefunction(handler):
+            self._async_handlers[event_type].append(handler)
+            subscription = Subscription(
+                lambda: self._async_handlers[event_type].remove(handler)
+            )
+        else:
+            self._sync_handlers[event_type].append(handler)
+            subscription = Subscription(
+                lambda: self._sync_handlers[event_type].remove(handler)
+            )
+
+        # notify subscriptions have been updated.
+        self._subscriptions_changed.set()
+
+        return subscription
+
+    TStreamEvent = TypeVar("TStreamEvent", bound=BaseEvent)
+
+    async def stream(
+        self, event_type: Type[TStreamEvent], num_events: Optional[int] = None
+    ) -> AsyncGenerator[TStreamEvent, None]:
+        """
+        Stream all events that match the specified event type. This returns an
+        ``AsyncIterable[BaseEvent]`` which can be consumed through an ``async for`` loop.
+        An optional ``num_events`` parameter can be passed to stop streaming after a maximum amount
+        of events was received.
+        """
+        (send_channel, receive_channel) = cast(
+            StreamChannelPair, trio.open_memory_channel(100)
+        )
```

Comment:
Huh, it's annoying that we have to add all these casts everywhere, it might be worth adding our own typestubs, or using someone elses: https://github.com/python-trio/trio-typing

Reply:
We're already using `trio-typing`.  I'm not sure it comes with anything for this but I'll look.